feat: includeMirrors

Rainman73 opened the pull request:
## Changes

Add support for 'includeMirrors' configuration option to allow Renovate to operate on (gitlab) repositories that are mirrors.

## Context

Closes #16436

## Documentation

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work 

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

Rainman73 commented:
As you can probably tell, I'm not a JS/TS programmer. I'm open to suggestions for more idiomatic solutions.

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -263,11 +264,13 @@ export async function initRepo({
   config.repository = urlEscape(repository);
   config.cloneSubmodules = cloneSubmodules;
   config.ignorePrAuthor = ignorePrAuthor;
+  config.includeMirrors = includeMirrors;
```
```suggestion
```
i don't think we need to store that option

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -280,7 +283,7 @@ export async function initRepo({
       throw new Error(REPOSITORY_ARCHIVED);
     }
-    if (res.body.mirror) {
+    if (res.body.mirror && !config.includeMirrors) {
```
```suggestion
    if (res.body.mirror && includeMirrors !== true) {
```
more explicit

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -72,6 +72,7 @@ let config: {
   cloneSubmodules: boolean | undefined;
   ignorePrAuthor: boolean | undefined;
   squash: boolean;
+  includeMirrors: boolean | undefined;
```
```suggestion
```
not needed, see below

viceice reviewed (commented):
mirror repos are still filtered on autodiscover, check `getRepos` function

Rainman73 commented:
> mirror repos are still filtered on autodiscover, check `getRepos` function



> mirror repos are still filtered on autodiscover, check `getRepos` function

Good point, the `.filter()` needs to be updated there, but how do I get the value of the `includeMirrors` setting there? (Sorry for the newb question)

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -1602,6 +1602,12 @@ If you need to force permanent unstable updates for a package, you can add a pac
 
 Also check out the `followTag` configuration option above if you wish Renovate to keep you pinned to a particular release tag.
 
+## includeMirrors
```
Move to self-hosted file

rarkins commented on lib/config/options/index.ts:
```diff
@@ -442,6 +442,14 @@ const options: RenovateOptions[] = [
+      'Whether to process repositories that are mirrors. By default, repositories that are mirrors are skipped.',
+    type: 'boolean',
+    default: false,
+    supportedPlatforms: ['gitlab'],
```
```suggestion
    supportedPlatforms: ['gitlab'],
    globalOnly: true,
```

rarkins reviewed (changes requested):
Make global only

rarkins commented:
Can you confirm that it works with --autodiscover flag?

Rainman73 commented:
> Can you confirm that it works with --autodiscover flag?

Actually I'm pretty sure it does *not* work with autodiscover because of the issue mentioned by viceice above. 

How do I get the value of the `includeMirrors` setting in the `getRepos` function in `lib/modules/platform/gitlab/index.ts` ?

rarkins commented:
@viceice I don't see any mirror filtering here: https://github.com/renovatebot/renovate/blob/81eb17aacac37dff6fa64def330f8fb0b428f4a3/lib/modules/platform/gitlab/index.ts#L146-L174

Rainman73 commented:
> @viceice I don't see any mirror filtering here:

Not viceice, but I think the filtering is being done on line 168:
https://github.com/renovatebot/renovate/blob/81eb17aacac37dff6fa64def330f8fb0b428f4a3/lib/modules/platform/gitlab/index.ts#L167-L169

rarkins commented:
Yes you're right, I missed that! So that line needs modification too

rarkins reviewed (changes requested):
Needs update to `getRepos()`

Rainman73 commented:
> Needs update to `getRepos()`

I updated `getRepos()` to use `GlobalConfig` and added a testcase, but when I set `includeMirrors` to true in the tests, the changes to GlobalConfig seem to be ignored. Should the `GlobalConfig` be mocked for the tests?

rarkins reviewed (changes requested):
Please modify `AutodiscoverConfig` instead of using GlobalConfig and then pass the param here: https://github.com/renovatebot/renovate/blob/440ea597fcab0fe687882f8170e72a57e4930f89/lib/workers/global/autodiscover.ts#L37-L40

Rainman73 commented:
As far as I can tell, I have made all requested changes, and the newly added tests now work correctly. Thanks for the patience!

rarkins commented on test/setup.ts:
```diff
@@ -151,7 +151,7 @@ declare global {
 
     interface Mock<T = any, Y extends any[] = any>
-      extends Function, // eslint-disable-line @typescript-eslint/ban-types
+      extends Function,
```
Unrelated change

Rainman73 replied:
In my (docker/vs code) dev env `yarn test` stumbles over this because of
 `154:25  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/ban-types')`

Am I doing something wrong, or should the eslint directive be reinstated?

rarkins replied:
Maybe you needed to run `yarn` again to get the latest dependencies

rarkins commented on docs/usage/self-hosted-configuration.md:
```diff
@@ -482,6 +482,12 @@ Use the `extends` field instead of this if, for example, you need the ability fo
 
+## includeMirrors
+
+By default, Renovate does not operate on repositories that are mirrors.
```
```suggestion
By default, Renovate does not autodiscover repositories that are mirrors.
```

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -267,6 +270,7 @@ export async function initRepo({
   ignorePrAuthor,
   gitUrl,
   endpoint,
+  includeMirrors,
```
```suggestion
```

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -284,7 +288,8 @@ export async function initRepo({
     }
-    if (res.body.mirror) {
+
+    if (res.body.mirror && includeMirrors !== true) {
```
@viceice What if we remove this if altogether?

1. If it's been autodiscovered then we know the includeMirrors flag must be true already
2. If the user has run it directly, they clearly want to include it

rarkins replied:
Let's do it later

rarkins commented:
Global only.

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -168,8 +169,10 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
       paginate: true,
     });
     logger.debug(`Discovered ${res.body.length} project(s)`);
+    const includeMirrors = config?.includeMirrors === true;
+
```
```suggestion
```

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -168,8 +169,10 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
+
     return res.body
-      .filter((repo) => !repo.mirror)
+      .filter((repo) => includeMirrors || !repo.mirror)
```
```suggestion
      .filter((repo) => config?.includeMirrors || !repo.mirror)
```

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -169,7 +170,7 @@ export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
     logger.debug(`Discovered ${res.body.length} project(s)`);
     return res.body
-      .filter((repo) => !repo.mirror)
+      .filter((repo) => config?.includeMirrors || !repo.mirror) // eslint-disable-line @typescript-eslint/prefer-nullish-coalescing
```
don't suppress this lint error

Rainman73 replied:
Without suppressing the lint error the build fails, which is why I used  
`const includeMirrors = config?.includeMirrors === true;` before, which @rarkins did not like.
What is the neatest way to fix this here?

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -72,6 +72,7 @@ let config: {
   cloneSubmodules: boolean | undefined;
   ignorePrAuthor: boolean | undefined;
   squash: boolean;
+  includeMirrors: boolean;
```
```suggestion
```
seems to be unused

Rainman73 commented on lib/config/global.ts:
```diff
@@ -28,6 +28,7 @@ export class GlobalConfig {
     'gitTimeout',
     'platform',
     'endpoint',
+    'includeMirrors',
```
You're right, I forgot to remove this when undoing the GlobalConfig changes

rarkins commented:
@Rainman73 please don't push from main branch unless you're making other changes. Each push resets the test approvals
